### ISSUE TITLE
Unable to import git repository into workspace

### ISSUE THREAD

**Issue by weeblr** [codenvy/codenvy#1839]
Hi

I cannot  import a (nearly empty ) test repository into a freshly, empty workspace. The repository:

- is hosted on a private Gitlab CE instance (latest as of 2017-02-27)
- the host has ssh access on a non-standard port (ie not 22)

**Reproduction Steps:**

- create a new workspace and run it
- using profile menu, create a new SSH key, and upload that to the chosen user profile in Gitlab
- Use the "Import project" menu of the workspace
- fill in Git contributor name and email address
- select "From Git URL" and enter the repo URL using this format:

````
git@gitlab.example.com:gitlab_group/gitlab_repo.git
````

Processing starts, but after maybe 30 seconds, the process times out and nothing has been imported

Notes:

- git cloning from the terminal, the same happen

This should be possible, at least against a gitlab.com repo, according to #628.

Is there any specific syntax for a non-standard SSH port? does the host needs to be whitelisted, as I gathered from various issues?

Rgds

**Codenvy version:**   5.3.1

**OS and version:**    irrelevant

**Docker version:** irrelevant

**Codenvy cli.log output:** irrelevant

**Comment by TylerJewell** [comment]
There was another issue identical to this one reported last night but I cannot find it now.  I think this may be to gitlab altering ip addresses and we need to ask our admins to verify the whitelist for gitlab.

**Comment by weeblr** [comment]
Hi

As mentioned above, this is not for Gitlab.com. This is a **private Gitlab instance**, running on our own domain.

I am not sure how/why you intend to have to whitelist all your users repos, but that seems like a very large task!

Rgds

**Comment by TylerJewell** [comment]
Sorry I didn't see that it was a private repo.  That should work over ssh without any additional configuration.  Will ask support to investigate.

We whitelist every provider on codenvy.io as a security prevention.  The effort is not large as most providers are just a couple of ip addresses.

**Comment by weeblr** [comment]
Ah OK, I thought you'd want to whitelist every remote git repo!

The only unusual thing here is the non-standard port, really.

Rgds

**Comment by TylerJewell** [comment]
@eivantsov @support - FYI

**Comment by eivantsov** [comment]
@weeblr can you share the IP of the node that hosts your GitLab server?

**Comment by weeblr** [comment]
@eivantsov I just sent you an email

**Comment by eivantsov** [comment]
@weeblr can you check it now please?

**Comment by weeblr** [comment]
@eivantsov Much better. A connection is happening now, but I do get "Connection refused", so I might have to work on the authorizations and stuff. I do have the ssh key from the profile/preferences/SSH/VCS dialog set up as an SSH key for a user with full access to that repo.
I assume in Codenvy the user used for conenction is selected by enterin name/email address under the Profile/preferences/Committer entry fields?

**Comment by eivantsov** [comment]
@weeblr how did you generate ssh key?

**Comment by weeblr** [comment]
I didn't. I am using the Codenvy one: https://monosnap.com/file/q0RDdjJYivuW74f9AKaDCAgcUgEHfr

**Comment by eivantsov** [comment]
Did you enter hostname without protocol? Did you save this key to your Gtlab account settings?

**Comment by weeblr** [comment]
Yes to both. The host is as per the email I sent you. And the ssh key was added under the "SSH keys" of the user (who has access to the repo on gitlab), which user name and email address I entered under the Profile/Preferences/Committer fields in Codenvy

**Comment by weeblr** [comment]
For the repo address (when in the Import project dialog), I use 

git@gitlab.example.com:gitlab_group/gitlab_repo.git

with 'gitlab.example.com" being the exact same host used for generating the SSH in Codenvy, and sent to you by email (but stripped of the scheme ofc)

**Comment by weeblr** [comment]
would the fact that this machine has SSH access over a non-standard port be a problem? (ie SSH is not on port 22)

**Comment by eivantsov** [comment]
Yes, that is the problem.

**Comment by weeblr** [comment]
Ah I mentioned it from the start, as I thought it could be an issue as you do not have a way to specify a port.

What do you suggest?

**Comment by eivantsov** [comment]
Sorry, I missed that.

Right now, there's no way to tell JGit that it should connect to your server on a non standard ssh port.

You may use native Git in the terminal.

**Comment by weeblr** [comment]
OK. Is there any advantage to using your GIT import feature over manual git in the terminal? Does making Codenvy aware of the existence of a GIT repo enables UI interaction (like fetch, push, switching branches) instead of having to use the terminal to perform those actions?

In other words, does the import function simply saves me a git clone command?

**Comment by eivantsov** [comment]
Yes, you can use git features, except for push operations

**Comment by weeblr** [comment]
Ah, so I have a choice between moving to port 22, or only using git from cli, right?

Still, I am puzzled that using a non-standard port is not possible. I hope you can manage that, it's kinda of a big drag.

Lastly, did you require to also whitelist my IP address to allow the connection? will it be like this for any new repo I would like to work with (on different servers)?

Rgds

**Comment by eivantsov** [comment]
If you use a different server or IP of the existing server changes, then yes, we have to whitelist this new IP.

And yes, either port 22 and using built in Git or native git in a container

**Comment by weeblr** [comment]
OK, thanks. 

1 more important comment though: you should make it very, very clear that you have to whitelist each and every server holding the remote repositories used.  This is a very large waste of time for multiple people, trying to make work something that can't. 

If too much to ask that you display a proper message in Codenvy, then at least make it a prominent thing in your documentation.

And of course, now would be a good idea to mention in the same place that you require those remote repo to be accessible only on port 22. Again, this is a very large waste of time for your potential customers.

**Comment by gazarenkov** [comment]
+1 of making it clear on documentation since I doubt we can change firewall's message.

**Comment by TylerJewell** [comment]
I think there are a few separate issues here.  And while documentation is always a help, I don't think that most users will get that far into the documentation to see the problem.

First, the primary issue is that we are a free service and do not require credit cards. We let you provision runtimes and with it being wide open this way, it's easy for malicious people to spin up a workspace and launch some bots illegally.  You then get into a game of whack-a-mole.  Requiring white listing of external providers is a crude technique for controlling the risk of this behavior.

Second, @weeblr - have you looked at running a custom Codenvy `docker run codenvy/cli start`? You can run this privately and no whitelisting required. There are also other configuration items that can be addressed.

Third @weeblr - I am not a git expert, but if you provide a `ssh://git@<url>:<ssh-port>/repo.git`, will this not tell git clients to use a non-standard ssh port? This syntax would be easier than going through some hoops to ask users to provide custom SSH ports for various hosts.

Fourth @ashumilova @eivantsov - the most common this place is going to happen is in the UD.  But the problem is that when you provide a custom git URL, you have to wait for a ws to be created, runtime started, agents started - and only then do we try to do a git clone.  If the git clone error happens because of a blocked host or a custom port, you will not really understand why you are getting an issue so late in the process.  So, it seems that we should add a host + port ping check in the UD for any git URL provided by a user *before* they ask to create a workspace.  We should do some sort of check mark positively next to any git URL entered to give the user confidence that it is reachable.  If it's not reachable, we should have some sort of error message and we can then provide a help tool tip to explain that you need to make sure your provider is white listed AND how to configure a non-standard SSH port.

Fifth, if you cannot do a non-standard SSH port, then we should extend our SSH key configuration in the user dashboard / IDE so that you can also associate a custom SSH port with each host.

**Comment by TylerJewell** [comment]
I've added this commit to the docs - https://github.com/codenvy/docs/commit/8c92a1c3e729cde23fa8b0f6b4208919c9ef577f

**Comment by weeblr** [comment]
Hi @TylerJewell, appreciate your detailed answer here.

1.  Agreed, there has to be some safeguards. I have tested a couple of services, which do not have this, but it does make sense to have some kind of control

2. We already run multiple such servers, and for our limited (as of now) use case, I'd rather do without having to run and maintain another server for that. I do appreciate that you make it possible though

3. I tried that initially, and failed, but this was before being whitelisted, so that would explain that. I just tried again, and Codenvy can indeed connect somehow, but that does not work actually, because the SSH users (required to gain access to the server) are not the same as the Gitlab server users. Meaning that when you do:

`git clone ssh://ssh_user@gitlab.example.com:NNN/gitlab_user/repo_name.git`

gitlab_user/repo_name.git is searched starting from user 'ssh_user' home directory, which causes the repo to not be found.

And of course, gitlab_user does not have SSH access to the machine, as this is not needed and not wanted for most users.

4. I'd agree that this is the place where I'd expect to see a warning about a/ whitelisting requirement and b/ standard port requirement

Documentation as well of course, but it's in "From Git URL" assistant those warning would be the most visible for sure.

For now, I guess we'll have to make do with using terminal git commands for all remote operations.

Thanks and regards
